### PR TITLE
Expose Ability to Configure Context's Intrinsic Objects

### PR DESCRIPTION
**Pull request by GrantMatejka**
Expose various QuickJS functions to set a context's intrinsic options. Adds support for these option in both runtimes and breaks out building of context pointer into helper function. Also updates various comments in respect to new intrinsic object support.

I am having trouble getting the test to pass as with unsupported intrinsics, the errors that arise make it hard to dispose of the handles

Closes https://github.com/justjake/quickjs-emscripten/issues/113

### CODE REVIEW

File: ts/runtime.ts [hunk 1]
```diff
@@ -160,6 +153,79 @@ export class QuickJSRuntime implements Disposable {
...
+      contextPointer = this.ffi.QTS_NewContextRaw(this.rt.value)
+      this.addContextIntrinsics(contextPointer, options.intrinsics)
+    } else {
+      contextPointer = this.ffi.QTS_NewContext(this.rt.value)
+    }
+
+    return contextPointer
+  }
+
+  private addContextIntrinsics(ctxPtr: JSContextPointer, intrinsics: Intrinsic[]): void {
```
Comment (justjake) [inline ts/runtime.ts:171]: Thanks for the PR. Can we avoid duplicating this s method? Why is it needed in both implementations?

File: ts/types.ts [hunk 2]
```diff
@@ -169,18 +174,18 @@ const DefaultIntrinsicsList = [
...
    * If unset, the default intrinsics will be used.
-   * To omit all intrinsics, pass an empty array.
+   *
+   * The base objects intrinsics will always be enabled,
+   *  but to omit all additional intrinsics, pass an empty array.
+   * 
+   * To be able to use `evalCode` you must enable the "Eval" intrinsic.
    */
-  intrinsics?: PartiallyImplemented<Intrinsic[]> | typeof DefaultIntrinsics
+  intrinsics?: Intrinsic[]
```
Comment (justjake) [inline ts/types.ts:188]: Ideally we can accept DefaultIntrinsics or similar if a user wants to add stuff on top of the defauls

File: ts/runtime.ts (hunk 1 above)
Comment (GrantMatejka) [reply ts/runtime.ts]: forsure. The method is the same across runtimes, I just forgot runtime-async extends runtime. Addressed in https://github.com/justjake/quickjs-emscripten/pull/116/commits/9836ad8e33b8020f5ee42d4d6fe00ba2953de614

File: ts/types.ts (hunk 2 above)
Comment (GrantMatejka) [reply ts/types.ts:188]: Hmm, I'm not sure the best way to support that. 
What if we export the default intrinsic list and the user can extend it on their side if desired?

File: ts/quickjs.test.ts [hunk 3]
```diff
@@ -679,6 +685,67 @@ function contextTests(getContext: () => Promise<QuickJSContext>, isDebug = false
...
+    it("evalCode - context executes as expected with default intrinsics", async () => {
+      // Eval is required to use `evalCode`
+      const newContext = await getContext({ intrinsics: [...DefaultIntrinsics] })
+
+      let handle;
+      try {
+        handle = newContext.unwrapResult(newContext.evalCode(`new Date()`))
+      } catch (e) {
+        // No error should occur
+        throw e
```
Comment (GrantMatejka) [inline ts/quickjs.test.ts:739]: @justjake When trying to add intrinsics to a context, an error gets throw here (`RuntimeError: memory access out of bounds`) from https://github.com/GrantMatejka/quickjs-emscripten/blob/26865efcebc281739fca84f8c3b368d2840106f4/ts/context.ts#L741

I'm not entirely sure what is causing this

File: ts/types.ts (hunk 2 above)
Comment (justjake) [reply ts/types.ts:186]: The initial implementation had a stand-in symbol DefaultIntrinsics which if present should add all the default intrinsics. Giving the array works too and is certainly easier to understand.

File: c/interface.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
+#include <math.h>  // For NAN
```
Comment (justjake) [inline c/interface.h:1]: I don't think this file needs to be committed, it should be auto-generated by `make`

File: ts/runtime-asyncify.ts [hunk 5]
```diff
@@ -64,6 +61,19 @@ export class QuickJSAsyncRuntime extends QuickJSRuntime {
     return context
   }
 
+  private getNewContextPointer(options: ContextOptions = {}): JSContextPointer {
```
Comment (justjake) [inline ts/runtime-asyncify.ts:64]: What's the difference from `getNewContextPointer` in the parent class?

File: ts/quickjs.test.ts (hunk 3 above)
Comment (justjake) [reply ts/quickjs.test.ts:739]: Does the stack trace for DEBUG build include any stacks on the C side? One of LLVM's sanitizers like address sanitizer might help point to the issue. Docs here: https://emscripten.org/docs/debugging/Sanitizers.html#address-sanitizer

File: ts/runtime-asyncify.ts (hunk 5 above)
Comment (GrantMatejka) [reply ts/runtime-asyncify.ts:64]: The parent class allows for recycling of an already existing context, which didn't seem to be an existing feature of the async runtime

File: ts/quickjs.test.ts (hunk 3 above)
Comment (GrantMatejka) [reply ts/quickjs.test.ts:739]: Here is the stack trace for the Debug build test
```
 DEBUG async module
         sync API
           intrisic objects
             evalCode - context executes as expected with default intrinsics:
     RuntimeError: memory access out of bounds
      at JS_CallInternal (wasm://wasm/005ac8be:wasm-function[112]:0x156dd)
      at JS_EvalFunctionInternal (wasm://wasm/005ac8be:wasm-function[317]:0x5f71a)
      at __JS_EvalInternal (wasm://wasm/005ac8be:wasm-function[135]:0x2c686)
      at JS_Eval (wasm://wasm/005ac8be:wasm-function[321]:0x61256)
      at QTS_Eval (wasm://wasm/005ac8be:wasm-function[72]:0xae29)
      at ret.<computed> (ts/generated/emscripten-module.WASM_DEBUG_ASYNCIFY.js:1646:35)
      at /home/grantmatejka/quickjs-emscripten/ts/generated/emscripten-module.WASM_DEBUG_ASYNCIFY.js:732:22
      at ccall (ts/generated/emscripten-module.WASM_DEBUG_ASYNCIFY.js:1895:22)
      at /home/grantmatejka/quickjs-emscripten/ts/generated/emscripten-module.WASM_DEBUG_ASYNCIFY.js:1934:16
      at QuickJSAsyncFFI.mustBeSync [as QTS_Eval] (ts/types-ffi.ts:108:20)
      at /home/grantmatejka/quickjs-emscripten/ts/context.ts:741:18
      at Lifetime.consume (ts/lifetime.ts:98:20)
      at QuickJSAsyncContext.evalCode (ts/context.ts:740:8)
      at Context.<anonymous> (ts/quickjs.test.ts:737:53)
```

Comment (GrantMatejka) [reply ts/quickjs.test.ts:739]: Here is the stack trace from the debug test case using default intrinsics
``` 
RuntimeError: memory access out of bounds
    at JS_GetGlobalVar (wasm://wasm/006b2e22:wasm-function[553]:0x5755d)
    at JS_CallInternal (wasm://wasm/006b2e22:wasm-function[405]:0x32940)
    at JS_CallFree (wasm://wasm/006b2e22:wasm-function[397]:0x2c739)
    at JS_EvalFunctionInternal (wasm://wasm/006b2e22:wasm-function[626]:0x68556)
    at __JS_EvalInternal (wasm://wasm/006b2e22:wasm-function[263]:0x18832)
    at JS_EvalInternal (wasm://wasm/006b2e22:wasm-function[631]:0x69c36)
    at JS_EvalThis (wasm://wasm/006b2e22:wasm-function[630]:0x69af7)
    at JS_Eval (wasm://wasm/006b2e22:wasm-function[632]:0x69cd8)
    at QTS_Eval (wasm://wasm/006b2e22:wasm-function[102]:0x6332)
    at /home/grantmatejka/quickjs-emscripten/ts/generated/emscripten-module.WASM_DEBUG_SYNC.js:728:22
    at ccall (/home/grantmatejka/quickjs-emscripten/ts/generated/emscripten-module.WASM_DEBUG_SYNC.js:4508:22)
    at QuickJSFFI.QTS_Eval (/home/grantmatejka/quickjs-emscripten/ts/generated/emscripten-module.WASM_DEBUG_SYNC.js:4525:16)
    at /home/grantmatejka/quickjs-emscripten/ts/context.ts:741:18
    at Lifetime.consume (/home/grantmatejka/quickjs-emscripten/ts/lifetime.ts:98:20)
    at QuickJSContext.evalCode (/home/grantmatejka/quickjs-emscripten/ts/context.ts:740:8)
    at Context.<anonymous> (/home/grantmatejka/quickjs-emscripten/ts/quickjs.test.ts:736:53)
Aborted(Assertion failed: list_empty(&rt->gc_obj_list), at: quickjs/quickjs.c,1983,JS_FreeRuntime)
        4) evalCode - context executes as expected with default intrinsics
```

Comment (GrantMatejka) [reply ts/quickjs.test.ts:739]: I'm pretty stumped on this and compiling with address sanitizer leads to a lot of `undefined symbol:` errors

Comment (GrantMatejka) [reply ts/quickjs.test.ts:739]: Disregard, but was that BaseObjects and BasicObjects are different 🤦 
BasicObjects are always attached to any context, but base objects need to be attached

**Comment by justjake** [comment]
Sorry about never merging your change. I did a big library shuffle recently, so your PR is stale. No worries! I'll take care of merging it here: #136

**Pull request closed without merging**